### PR TITLE
Perform git-related tests in different repo

### PR DESCRIPTION
**Pull request by thumbnail**
## Brief

Suggestion as discussed in #167.
 

## QA plan

1. run tests
2. change some files
3. run tests again

Note no test failure related to git-status not being pristine

## Author checklist

* [x] I have QAed the functionality
* [x] The PR has a reasonably reviewable size and a meaningful commit history
* [x] I have run the [branch formatter](https://github.com/nedap/formatting-stack/blob/332a419034ab46fad526a5592f4257353bd695b6/src/formatting_stack/branch_formatter.clj) and observed no new/significative warnings
* [x] The build passes
* [x] I have self-reviewed the PR prior to assignment
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [x] Correctness
  * [x] Robustness (red paths, failure handling etc)
  * [x] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [x] Security
  * [x] Performance
  * [x] Breaking API changes
  * [x] Cross-compatibility (Clojure/ClojureScript)

## Reviewer checklist

* [ ] I have checked out this branch and reviewed it locally, running it
* [ ] I have QAed the functionality
* [ ] I have reviewed the PR
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [ ] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [ ] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [ ] Breaking API changes
  * [ ] Cross-compatibility (Clojure/ClojureScript)

### CODE REVIEW

File: test/integration/formatting_stack/strategies.clj [hunk 1]
```diff
@@ -49,7 +61,7 @@
       (-> deletable-file .delete)
       (expect-sane-output! (sut/all-files :files []))
       (finally
-        (sh "git" "checkout" (str deletable-file))
+        (sh "git" "checkout" ".")
```
Comment (thumbnail) [inline test/integration/formatting_stack/strategies.clj:64]: Other tests already relied on `.`, since we run these commands in a copied (temporary) repo **and** assert `git pristine` before, i think this is fine.

File: src/formatting_stack/strategies/impl.clj [hunk 2]
```diff
@@ -27,7 +31,9 @@
 
 (speced/defn ^::file-entries file-entries
   [& args]
-  (->> args (apply sh) :out string/split-lines (filter seq)))
+  ;;; xxx
+  (let [r (apply sh (conj (vec args) :dir *pwd*))]
+    (->> r :out string/split-lines (filter seq))))
```
Comment (thumbnail) [inline src/formatting_stack/strategies/impl.clj:36]: There should be a better way to add the `:dir` option here.

File: test/integration/formatting_stack/strategies.clj (hunk 1 above)
Comment (vemv) [reply test/integration/formatting_stack/strategies.clj:64]: I don't know, it would be very frustrating if some programming error meant that one's WIP was erased.

**Review by vemv — COMMENTED** [review]
The approach seems sound to me, but I'd make sure that no WIP is possibly lost

File: test/integration/formatting_stack/strategies.clj (hunk 1 above)
Comment (thumbnail) [reply test/integration/formatting_stack/strategies.clj:64]: True, but I don't think it's likely. However; simply introducing another constant here should be doable.

File: .gitignore [hunk 3]
```diff
@@ -28,3 +28,4 @@ pom.xml
 pom.xml.asc
 profiles.clj
 repl-port
+/tmp
```
Comment (vemv) [inline .gitignore:31]: maybe we can use `git-integration-testing`, so that the intent is clearer and don't end up storing unrelated stuff under `tmp`

File: test/integration/formatting_stack/strategies.clj [hunk 4]
```diff
@@ -1,28 +1,53 @@
...
 
-(def ^String createable-filename "test-resources/createable.clj")
+(def ^File rename-destination (File. "tmp/test-repo/test-resources/deletable_moved.clj"))
+
+(def ^String creatable-filename "tmp/test-repo/test-resources/creatable.clj")
 
 (def creatable-contents (pr-str '(ns foo)))
 
-(assert (-> deletable-file .exists))
+(defn cleanup-repo []
```
Comment (vemv) [inline test/integration/formatting_stack/strategies.clj:27]: maybe can be named `cleanup-testing-repo!` so that it's clearer that the main repo is not intended to be cleaned up

File: src/formatting_stack/strategies/impl.clj (hunk 2 above)
Comment (thumbnail) [reply src/formatting_stack/strategies/impl.clj]: I implemented a rudimentary approach where I use 1 impl for the `git`-command here; 15ba8e1

That suggestion also removes the `git-command`-def, and makes sure any git-command is executed in the right pwd. It removes /some/ flexibility from `impl/absolutize`, but we can extend/restore that if we ever need it in the future.

WDYT @vemv ?

File: src/formatting_stack/strategies/impl.clj [hunk 5]
```diff
@@ -38,13 +48,12 @@
...
                                              (-> s File. .exists)))))
 
-(speced/defn ^::existing-files absolutize [command, ^::file-entries file-entries]
-  (let [toplevel-fragments (case command
-                             "git" (-> (sh "git" "rev-parse" "--show-toplevel")
-                                       (:out)
-                                       (string/split #"\n")
-                                       (first)
-                                       (string/split separator-pattern)))]
+(speced/defn ^::existing-files absolutize [^::file-entries file-entries]
```
Comment (vemv) [inline src/formatting_stack/strategies/impl.clj:51]: `strategies/impl` is git-agnostic (`"git"` is only ever handled over a `case`), while `stategies.clj` makes git part of its explicit api (e.g. a given defn is called `git-completely-staged`, not `completely-staged`).

So as of `master`, we're very clear about what we promise / intend to offer: an opt-in git interface, and some agnostic implementation details.

I wouldn't deviate from that as f-s, should be usable in non-git environments (e.g. some enterprises).

---

(speaking of, we can make that `case` a multimethod instead - only an _open_ set is truly agnostic)

Comment (thumbnail) [reply src/formatting_stack/strategies/impl.clj:51]: As eluded to here: https://github.com/nedap/formatting-stack/pull/185#discussion_r601425939; This change might look like flexibility is removed, but the flexibility is a false promise. The function is not open for extension, and only requires any consumer to provide a static argument (`"git"` in this case). Consumers without git would need their own replacements for the strategies anyway. 

This is an internal function anyway; which is subject to not public by definition. While I agree with the sentiment it's a opt-in git interface, i don't agree that this change affects that in any way.

Comment (vemv) [reply src/formatting_stack/strategies/impl.clj:51]: >  but the flexibility is a false promise. 

It trivially becomes a true promise by refactoring a single `case` to a multimethod. We can be thankful for having discovered that via this PR.

> While I agree with the sentiment it's a opt-in git interface, i don't agree that this change affects that in any way.

It does increase the cost of implementing non-git strategies. As of this branch, if tomorrow someone created a PR with `hg` support, we'd plausibly have a hard time adapting internals, tests, etc because everything was more coupled to `git` than before.

Comment (thumbnail) [reply src/formatting_stack/strategies/impl.clj:51]: > It trivially becomes a true promise by refactoring a single case to a multimethod. We can be thankful for having discovered that via this PR.

I don't think it's something we should pursue. The current naming of te strategies are not abstracted properly from git. Mecurial for example, doesn't have any notion of 'staged' files. Refactoring the case to a multimethod does not fix that.

> It does increase the cost of implementing non-git strategies.

I'm not sure about this. I prefer to have small, specific namespaces. If another vcs would be supported i would not put it in `formatting-stack.strategies`, but in a more specific one. Especially if you keep the opt-in promise in mind. 

I don't see how the amount of work would be affected.

Comment (vemv) [reply src/formatting_stack/strategies/impl.clj:51]: Ok but let's rename absolutize -> git-absolutize

Also https://github.com/nedap/formatting-stack/pull/184 should be merged before this one, since that one precisely adds coverage for this sort of things

Comment (thumbnail) [reply src/formatting_stack/strategies/impl.clj]: This change was mostly driven from the necessity to run the `git` command in a specific directory for both the test-suite and the code. Which in turn, presented the opportunity remove the `"git"`-definition and dependent code. This is extracted into 
 e61c276794575714f3015c489f8b8e241f8b204


I discovered `clojure.java.shell/with-sh-dir`, which perfectly suits our usecase here, so i undid related changes.

**Comment by thumbnail** [comment]
@neominik I implemented your `git clone` suggestion, which worked like a charm. If you feel like a review, go for it !

File: test/integration/formatting_stack/strategies.clj [hunk 6]
```diff
@@ -3,26 +3,46 @@
...
+   [clojure.test :refer [deftest is testing use-fixtures]]
    [formatting-stack.strategies :as sut]
+   [formatting-stack.test-helpers :as test-helpers]
    [formatting-stack.util :refer [rcomp]]
    [nedap.speced.def :as speced])
   (:import
    (java.io File)))
 
-(def ^File deletable-file (File. "test-resources/deletable.clj"))
+(def ^File deletable-file (File. "git-integration-testing/test-resources/deletable.clj"))
```
Comment (vemv) [inline test/integration/formatting_stack/strategies.clj:14]: You can use `io/file` to avoid having `/`s around

File: test/integration/formatting_stack/strategies.clj [hunk 7]
```diff
@@ -3,26 +3,46 @@
...
 
-(def ^File rename-destination (File. "test-resources/deletable_moved.clj"))
+(def ^File rename-destination (File. "git-integration-testing/test-resources/deletable_moved.clj"))
```
Comment (vemv) [inline test/integration/formatting_stack/strategies.clj:16]: Would you use `"git-integration-testing"` as an extracted constant?

File: test/integration/formatting_stack/strategies.clj [hunk 8]
```diff
@@ -3,26 +3,46 @@
...
+  (fn [t]
+    (assert (-> deletable-file .exists))
+
+    (assert (not (-> rename-destination .exists)))
+
+    (assert (not (-> creatable-filename .exists)))
+
+    (t)))
 
 (defn assert-pristine-git-status! []
```
Comment (vemv) [inline test/integration/formatting_stack/strategies.clj:47]: Is `assert-pristine-git-status!` still useful?

(Might be part of the QA plan)

**Review by vemv — APPROVED** [review]
LGTM!

File: test/integration/formatting_stack/strategies.clj (hunk 8 above)
Comment (thumbnail) [reply test/integration/formatting_stack/strategies.clj:47]: While developing this PR did catch some nasty mistakes, so i'm happy to keep it in tbh.

Comment (vemv) [reply test/integration/formatting_stack/strategies.clj:47]: SGTM, I just wasn't sure (that is, without running it) whether it was still a runing a valid assertion

**Review by josefigueroa-nedap — APPROVED** [review]
Tested and monitored during execution. It works pretty well! 👍
